cpp/src/parquet/encoding.cc

```diff
@@ -127,6 +130,25 @@ class PlainEncoder : public EncoderImpl, virtual public TypedEncoder<DType> {
   }
 
  protected:
+  template <typename ArrayType>
+  void PutBinaryArray(const ArrayType& array) {
+    const int64_t total_bytes =
+        array.value_offset(array.length()) - array.value_offset(0);
+    PARQUET_THROW_NOT_OK(sink_.Reserve(total_bytes + array.length() * sizeof(uint32_t)));
+
+    constexpr size_t kMaxByteArraySize = std::numeric_limits<uint32_t>::max();
```

Comment:
its not clear from the spec, but I think this expected to be signed:

[Java](https://github.com/apache/parquet-mr/blob/7469e87dc63c0f672be7175946dc7af1219ec732/parquet-column/src/main/java/org/apache/parquet/column/values/plain/BinaryPlainValuesReader.java#L39) appears to use a int (signed 32 bit integer)

Reply:
I've changed the code to check for the signed maximum. It will be safer in any case.